content/blog/2023/2024-01-27-deserialize-with-jsoncreator.md

```diff
@@ -0,0 +1,465 @@
+        this.lastName = map.get("lastName");
+        this.city = map.get("city");
+    }
+````
+When we pass a serialized Map object to the `ObjectMapper` class, it will automatically make use of the DELEGATING mode to create the `UserData` object.
+````java
+    public static Map<String, String> getMockedUserDataMap() {
+        return Map.of("id", "100", "firstName","Ranjani", "lastName","Harish", 
+        "city","Sydney");
+    }
+        
+    @Test
+    public void jsonCreatorWithDelegatingMode3() throws JsonProcessingException {
+        String userDataJson = 
+            objectMapper.writeValueAsString(getMockedUserDataMap());
+        assertNotNull(userDataJson);
+        UserData data = objectMapper.readValue(userDataJson, UserData.class);
+    }
+````
+
+Now that we understand how to use the `@JsonCreator` annotation in Java, let's look at a specific use case where this annotation is required in a Spring Boot application.
+Let's create a basic Spring Boot application with Rest endpoints that support pagination. **Pagination is the concept of dividing a large number of records in parts/slices. It is particularly useful, when we have to create REST endpoints to be consumed by a front end application.
+The Spring paging framework converts this data for us which makes retrieving data easier.**
+This sample application only demonstrates the usage of `@JsonCreator`. To understand how pagination works in Spring Boot, refer [this article.](https://reflectoring.io/spring-boot-paging/)
+
+This User Application uses H2 database to store and retrieve data. 
+The application is configured to run on port 8083, so let's start our application first :
+````text
+mvnw clean verify spring-boot:run (for Windows)
+./mvnw clean verify spring-boot:run (for Linux)
+````
+
+To demonstrate pagination, we have converted the `List<User>` returned from the DB into a paged object as shown below :
+````java
+    @GetMapping("/userdetails/page")
+    public ResponseEntity<Page<UserData>> getPagedUser(
+                    @RequestParam(defaultValue = "0") int page,
+                    @RequestParam(defaultValue = "20") int size) {
+        List<UserData> usersList = userService.getUsers();
+
+        // First let's split the List depending on the pagesize
+        int totalCount = usersList.size();
+        int startIndex = page * size;
+        int endIndex = Math.min(startIndex + size, totalCount);
+
+        List<UserData> pageContent = usersList.subList(startIndex, endIndex);
+
+        Page<UserData> employeeDtos = 
+                new PageImpl<>(pageContent, PageRequest.of(page, size), totalCount);
+
+        return ResponseEntity.ok()
+                .body(employeeDtos);
+    }
+````
+Here, the `Page` class is an interface in the `org.springframework.data.domain` package.
+When we make a GET request to the endpoint `http://localhost:8083/data/userdetails/page`, we see the JSON response as below:
+````json
+{
+    "content": [
```

Comment:
The response here already has 2 elements only. This was addressed as part of review comments earlier. The rest of the JSON is the pagination metadata. I wanted to keep it as is since the scenario talks specifically about pagination, so it would be helpful to the readers to understand how the pagination metadata looks like. Let me know if that makes sense to you.